### ISSUE TITLE
如何实现解析到第一个I帧立刻开始渲染?

### ISSUE THREAD

**Issue by dourgulf** [Bilibili/ijkplayer#1521]
最近看了一下关于直播秒开的文章(比如http://url.cn/2LAhROU)
其中关于播放器的优化都提到一个点"大部分播放器都是拿到一个完整的 GOP 后才能解码播放， 改写播放器逻辑让播放器拿到第一个关键帧（I帧）后就给予显示。不需要等待全部的GOP下载完毕才开始解码；"

请问一下 @bbcallen ijkplayer现在是拿到完整的GOP后才播放吗? 如果是, 应该在什么地方下手可以改成拿到关键帧就显示?

**Comment by bbcallen** [comment]
See avformat_find_stream_info()

**Comment by dourgulf** [comment]
I will try, share later.

**Comment by xingzhao-tt** [comment]
@dourgulf 

His Dourgulf,

Have you got the answer? If yes, could you help share it? :)

Thanks,
Felix

**Comment by dourgulf** [comment]
@xingzhao-tt 努力啃ffmpeg代码中:)

**Comment by hondameng** [comment]
@dourgulf I am waiting for your good news,if you solve it ,please @me,thanks

**Comment by dourgulf** [comment]
@hondameng 最近找到了靠谱的CDN, 发现ijkplayer不用改, 已经可以秒开(平均0.6s打开). 另外一家合作的CDN就一直在avformat_find_stream_info卡主1.5s以上. 暂时就这这家好的CDN用着, 不在深究ijkplayer了~, Issue留着以后继续回来.

**Comment by Android4MediaPlayer** [comment]
能分享测试地址吗

**Comment by dourgulf** [comment]
@Android4MediaPlayer 抱歉, 公司商业用途的服务器, 无法分享

**Comment by hondameng** [comment]
@dourgulf 我用的服务器，如果是用ijkplayer的话延迟达到6s左右，然后我用别的拉流播放SDK进行播放，延迟1s不到，是什么原因呢

**Comment by zhaoxiaobao** [comment]
延迟问题，解决了没呀？ @hondameng

**Comment by hondameng** [comment]
@zhaoxiaobao not yet

**Comment by limbo0312** [comment]
@dourgulf  这位同学，fork了一个探讨秒开的分支在这里是吧？
 [https://github.com/dourgulf/ijkplayer](url)

描述的很好。如果能够留下后面你们的解决办法就更好了 😄

**Comment by yangkai3** [comment]
@dourgulf 你公司用的哪家cdn？求一下

**Comment by linjiansheng** [comment]
发现ijkplay 已经发通知 IJKMPMoviePlayBackStateDidChange 1: playing
但是继续读取n个包，需要确认下继续读取n个包是不是为了满足读取一个完整的GOP如 @dourgulf 所描述的  ，才打印Video: first frame decoded。

**Comment by linjiansheng** [comment]
@dourgulf 我看你fork 里面设置了
ic->max_analyze_duration = 2000000;
ic->flags |= AVFMT_FLAG_NOBUFFER;  // 这个东东会导致max_analyze_duration预分析读取到的数据直接扔掉，而不会给解码器，加入GOP = 2，那么一开始读取的2秒被扔弃了，继续读一个完整的GOP，也就是要当dts = 4的时候，才解码出数据来
@bbcallen

**Comment by dourgulf** [comment]
是的, 参数调整的方案并不理想, 因此后续也没有再这个思路上继续往下走了; 现在主要精力改成直接做一个针对RTMP的播放器更实在. ijkplayer的目标太大, 不适合直播的播放优化体验.

**Comment by xxycode** [comment]
@dourgulf 可以试一下PLPlayerKit

**Comment by dourgulf** [comment]
PLPlayerKit效果挺不错的；遗憾的是它闭源的（我考察的时候是闭源）。而且， 有几个核心需求提了issue也没有人理会， 于是放弃了（我当时的几个需求是，首个画面的渲染事件， 视频大小改变的回调事件）

**Comment by lushan1314** [comment]
cdn用网宿。
